rsciio/utils/distributed.py

```diff
@@ -0,0 +1,169 @@
+#
+# RosettaSciIO is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# RosettaSciIO is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with RosettaSciIO. If not, see <https://www.gnu.org/licenses/#GPL>.
+
+
+import numpy as np
+import dask.array as da
+
+
+def get_chunk_slice(
+    shape,
+    chunks="auto",
+    block_size_limit=None,
+    dtype=None,
+):
+    """
+    Get chunk slices for the :func:`rsciio.utils.distributed.slice_memmap` function.
+
+    Takes a shape and chunks and returns a dask array of the slices to be used with the
+    :func:`rsciio.utils.distributed.slice_memmap` function. This is useful for loading data from a memmaped file in a
+    distributed manner.
+
+    Parameters
+    ----------
+    shape : tuple
+        Shape of the data.
+    chunks : tuple, optional
+        Chunk shape. The default is "auto".
+    block_size_limit : int, optional
+        Maximum size of a block in bytes. The default is None.
+    dtype : numpy.dtype, optional
+        Data type. The default is None.
+
+    Returns
+    -------
+    dask.array.Array
+        Dask array of the slices.
+    """
+
+    chunks = da.core.normalize_chunks(
+        chunks=chunks, shape=shape, limit=block_size_limit, dtype=dtype
+    )
+    chunks_shape = tuple([len(c) for c in chunks])
+    slices = np.empty(
+        shape=chunks_shape
+        + (
+            len(chunks_shape),
+            2,
+        ),
+        dtype=int,
+    )
+    for ind in np.ndindex(chunks_shape):
+        current_chunk = [chunk[i] for i, chunk in zip(ind, chunks)]
+        starts = [int(np.sum(chunk[:i])) for i, chunk in zip(ind, chunks)]
+        stops = [s + c for s, c in zip(starts, current_chunk)]
+        slices[ind] = [[start, stop] for start, stop in zip(starts, stops)]
+    return da.from_array(slices, chunks=(1,) * len(shape) + slices.shape[-2:]), chunks
+
+
+def slice_memmap(sl, file, dtypes, shape, **kwargs):
+    """
+    Slice a memmaped file using a tuple of slices.
+
+    This is useful for loading data from a memmaped file in a distributed manner. This takes
+    a slice of the dimensions of the data and returns the data from the memmaped file sliced.
+
+    Parameters
+    ----------
+    sl : array-like
+        An array of the slices to use. The dimensions of the array should be
+        (n,2) where n is the number of dimensions of the data. The first column
+        is the start of the slice and the second column is the stop of the slice.
+    file : str
+        Path to the file.
+    dtypes : numpy.dtype
```

Comment:
Why is `dtypes` plural?

Reply:
It's a result of the `map_blocks` function also having a dtype parameter which causes some problems.

Reply:
In most cases yes but I think that the `map_blocks` function pops off the `dtype` argument.  So if you want to specify a `dtype` you have to give it a different name otherwise dask will eat that kwarg. 

There are also cases where you might have a different dtype. For example some data is saved with an array and then a timestamp and then another array.  In these cases it might be better to have a custom dtype which accounts for these breaks.  I was going to add that to this function eventually.

Reply:
In any case the user shouldn't really be using the `slice_memmap` function directly and should instead be using the `memmap_distributed` function which copies the syntax for the `memmap` function.

Reply:
Yes, you are right! 👍